### ISSUE TITLE
Facing is set when using a BT mouse on macOS

### ISSUE THREAD

**Issue by Azhrei** [RPTools/maptool#413]
**Describe the bug**
Zooming in/out using the Apple BT mouse on macOS can set the facing on a selected token if that token is owned by the player.

**To Reproduce**
1. Start MapTool.
2. Drop any token on the Grasslands map.
3. With the token selected, zoom in/out with the mouse.
4. Within a few attempts, facing will appear on the token.

**Expected behavior**
I _expect_ that zooming with the mouse does not set/change facing on the selected token.  Mouse handling on macOS has always been slightly buggered.  It appears to have something to do with the Ctrl key being used to implement right-click.

**MapTool Info**
- Version: 1.5.1
- Install: New (from .DMG)

**Desktop (please complete the following information):**
 - OS: macOS
 - Version 10.14.3 (but has been a problem since early 1.4)

**Additional context**
I have a small Java app that prints the contents of mouse events as they occur.  I will add some explanatory text to the output and attach it in a separate post.

**Comment by mrkwnzl** [comment]
The mouse wheel command for facing is left right. As Apple Mouses don‘t have a traditional mouse wheel, mouse wheel up/down also often triggers left/right.

My proposal: Create a toggle in the preferences which deactivates the behavior of setting the facing using mouse wheel left/right (there are two more ways to set the facing).

**Comment by Azhrei** [comment]
Closed for discussion in #417 instead.

**Comment by mrkwnzl** [comment]
I know that this was closed in favor of #417, but I think the issue of accidental facing is actually more pressing than moving around the map by scrolling with the mouse wheel.

I’m currently preparing a new game and try to prepare cool maps. Thing is, facing is also used as rotating images and other tokens. What happens is that I try to add some objects to a map (say some treasure chests or a fire place and beds). Since the facing also rotates the tokens, it’s really difficult to move around the map without destroying what I was doing. 

For example, I have a battle map of a forest as a background image. On top of that, I put some objects, like a fire place. I now want to move around the map or zoom out. Since setting the facing is so sensible, I rotate the fire place wildly. So what I need to do is to deselect the fire place token. Since it’s on top of the background map and I was zoomed in, I now accidentally selected the background map instead. When I now try to zoom out, I rotate the background image wildly.

So here’s my proposition (again, actually): Include a setting in the general tab to simply deactivate setting the facing with the mouse wheel. There are still two other ways to set the facing (keystrokes and right-click), so there wouldn’t be a problem with that.

Would that be feasible until there is a more elaborate solution which also solves #417?

**Comment by Azhrei** [comment]
I've done some research and I'm reopening this issue.  As with most things Apple, it's complicated.

First, some tests on my laptop show that the mouse wheel is generating "standard" wheel events, but with fields that don't always make sense.  Some background:  Many mice with wheels have "clicks" that can be felt while rotating the wheel.  The Java MouseWheelEvent is designed to report the number of clicks, either positive or negative, to indicate the amount of scroll desired by the user.  However, the Apple mouse is what Java would call a "high precision mouse wheel", meaning that an alternate field of the event is used to hold the `preciseWheelRotation` value.  That auxiliary field is `double`, while the "clicks" field is `int`.  You can probably see the issue:  the Apple mouse sends dozens (hundreds?) of mouse events in which the double precision value is less than 1, yet MapTool only uses the "clicks" field and thus only ever sees a 0 for positive scroll events.

To make matters worse, each place in the code that MapTool is handling mouse wheel events has its own implementation of how to track wheel movement.  Some places in the code use `e.getWheelRotation()` to retrieve the number of clicks and check whether it is less than 0, while other places retrieve it and check to see whether it is greater than 0.  This means that positive scrolling can never work in some parts of the code.

I'm considering creating a separate MouseWheelAdapter class that all mouse wheel handlers will delegate to.  However, this is a pretty big change that will require a lot of testing.  I'm going to be designing how this will work over the next few days, but I likely won't get to coding soon.  (I'm pointing that out in case another dev with an Apple mouse wants to take this on before I can get to it.  However, this affects me _every time I use MapTool_ so it's high priority for me.)

**Comment by Phergus** [comment]
What I want to know is why the facing is changing when the intended behavior is that only happens if the shift key is held down?  Why does MT think the shift key is down on the Mac?

**Comment by mrkwnzl** [comment]
I don’t think MapTools thinks that the shift key is down, at least not all the time. All other instances in which I need to press the shift key work as expected (such as set the facing using keystrokes or when displaying the stat sheet only when shift is pressed).

Plus, I’ve had a player accidentally change the facing on a windows machine. I’m not entirely sure what he did, but I think it’s unlikely that he pressed the shift key while zooming. Are you sure that the shift key is needed, and not only a right or left mouse wheel click? I don’t have a windows machine, so I can’t test it, but I’ll ask my players to try it when we play next.

**Comment by Azhrei** [comment]
Apparently, because it is.

It looks like on OSX, we'll have to just filter any wheel events that have Shift and ignore them.  But that means we couldn't use Shift and a "normal" _vertical_ wheel movement.  I'll look through all of the existing code that uses the wheel and see if we can change it to Alt, or just disable it on OSX and let the user use the keyboard to change facing.


Some diagnostics, reported here for posterity:

Here's an event from a test app I wrote.  This is for scrolling left using the wheel:
```
MOUSE_WHEEL,(242,132),absolute(2400,1178),button=0,modifiers=⇧,extModifiers=⇧,clickCount=0,scrollType=WHEEL_UNIT_SCROLL,scrollAmount=1,wheelRotation=0,preciseWheelRotation=0.1
```
You'll notice that the `extModifiers` field (the one used by MT) has the Shift button turned on.  Scrolling to the right also has the Shift turned on, but `preciseWheelRotation` is a negative number.  (These strings were generated by calling `ev.paramString()`, which apparently embeds the symbol for Shift instead of the word "shift".  Somebody decided to get cute, I guess.)

It appears that Java on Windows registers for `MOUSE_WHEEL` and not `MOUSE_HWHEEL`, so MapTool will never receive horizontal wheel events.  And Linux translates horizontal wheel events into `BUTTON_4` and `BUTTON_5`.

On macOS, when an application registers for wheel events, **all** wheel events are generated.  As you can see, the Shift key is turned on for horizontal events.  My research turned up some games that have this issue, like MineCraft.  However, MineCraft had used a library (LWJGL) that filtered horizontal events in v2, but stopped filtering in v3, so the MC folks had to add application-level filtering on their own.

**Comment by Phergus** [comment]
Wow.  Why report the shift key as down if it isn't?   Seems clear that separate mouse handlers for each OS are needed.

**Comment by Azhrei** [comment]
My guess is that they were trying to fit the data into the available range of predefined values for the fields.

But I agree — I hate it when a driver sets bits of predefined fields and uses them to mean something else.  Unix sort of does that with the st_mode field of the inode and the O_RDONLY/WRONLY/RDWR flags to open().  It’s messed up.

**Comment by Phergus** [comment]
Having a pref setting for disabling facing changes via mouse wheel would provide a solution.

**Comment by Azhrei** [comment]
I've written up a proposal for a solution that allows the horizontal wheel scrolling to work on macOS.  It's part of the wiki found [in my fork of MapTool](https://github.com/Azhrei/maptool/wiki/Centralized-mouse-wheel-handler).

I consider the short term solution very low risk and easy to implement.  It could be merged immediately.  The long term solution is not much more effort, but carries slightly more risk and could be queued for later merging.

**Comment by Phergus** [comment]
No adverse affects seen on Windows.  Will need Mac testing.

**Comment by mrkwnzl** [comment]
I played around with 1.5.11 RC 1. Everything seems to be working (from my end-user perspective) on macOS. Where I need the shift key, it works as expected. I no longer change the facing accidentally. Both other means of setting the facing work.

Unexpected, but not a pressing issue, is that scrolling left or right also zooms, which leads to conflicting behavior when I scroll diagonally (right and up, or left and down). I just mention this for completeness sake, I don’t expect this to actually hinder me in any way. The immediate problem (accidentally changing the facing) is resolved for me.

Thanks a lot for this!

**Comment by Phergus** [comment]
That's great!

@Azhrei - Do you want to look at the side scrolling issue?

**Comment by Azhrei** [comment]
Yes, I do!

I see what’s happening.  The high precision mouse is triggering horizontal scroll events when the user lifts or moves their finger even just slightly left or right.  It generates fractional movement amounts, but if there had already been a small amount built up during vertical scrolling — and the horizontal lift/move happens soon enough — it’ll trigger the horizontal.  With the Shift key removed, it appears to be a vertical scroll, ie. a zoom event.

This is related to my note, above, about a large vertical scroll that invokes inertial scrolling and when the user then presses/releases the Shift key, the inertia is still active but the wheel events change from vertical to/from horizontal.

I’m not sure what can be done about this right now, but I’ll look into it.

**Comment by Phergus** [comment]
Okay.  Will consider this closed for 1.5.11.  Can open a new issue for side scrolling zoom issue.
